In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datasets import load_dataset

In [2]:
dataset = pd.read_csv('zno_translated.csv')

/var/folders/cz/wq9d8j_11fx3b8pjjbk7z8br0000gn/T/ipykernel_55856/2814263813.py:1: DtypeWarning: Columns (83,84,85,86,87,88,102) have mixed types. Specify dtype option on import or set low_memory=False.
  dataset = pd.read_csv('zno_translated.csv')


In [3]:
dataset.head()

outid  birth sextypename   
0  a4e039d0-07ef-4a3d-820d-106dcaf01b96   2001        Male  \
1  c1fa1ed0-d2af-4964-9a88-3c138008304b   2001      Female   
2  6504a7f3-7dfd-4a64-a2d9-4d49f5021303   2000      Female   
3  5cbcd343-2e63-48b8-b36d-48797a76be31   2000        Male   
4  063c6498-7345-492c-b567-174938bac671   1999        Male   

              regname             areaname   
0  Zakarpattia Oblast     Berehiv district  \
1         Kyiv Oblast          Kyiv region   
2      Kharkiv Oblast              Kharkiv   
3        Rivne Oblast  Mlynivskyi district   
4        Rivne Oblast         Rivne region   

                              tername   
0                the village of Koson  \
1                             Brovary   
2  Chervonozavod district of the city   
3               the village of Mlyniv   
4                              Ostrog   

                                         regtypename   
0  A graduate of an Ukrainian school of the curre...  \
1  A graduate of an Ukrainian school of the curre...   
2  A graduate of an Ukrainian school of the curre...   
3  A graduate of an Ukrainian school of the curre...   
4                                   College graduate   

                                              eoname   
0  Kosonsk comprehensive school of grades I-III w...  \
1  Brovary specialized school of grades I-III #7 ...   
2  Kharkiv Gymnasium No. 12 of Kharkiv City Counc...   
3  Mlyniv secondary school of grades I-III #1 of ...   
4  State Vocational and Technical Educational Ins...   

                           eotypename           eoregname  ... engdpalevel   
0  Secondary general education school  Zakarpattia Oblast  ...         NaN  \
1                  Specialized school         Kyiv Oblast  ...     Profile   
2                           Gymnasium      Kharkiv Oblast  ...     Profile   
3  Secondary general education school        Rivne Oblast  ...         NaN   
4            Higher vocational school        Rivne Oblast  ...         NaN   

  engball fradpalevel fraball deudpalevel deuball spadpalevel spaball   
0     NaN         NaN     NaN         NaN     NaN         NaN     NaN  \
1    63.0         NaN     NaN         NaN     NaN         NaN     NaN   
2    55.0         NaN     NaN         NaN     NaN         NaN     NaN   
3     NaN         NaN     NaN         NaN     NaN         NaN     NaN   
4     NaN         NaN     NaN         NaN     NaN         NaN     NaN   

  ukradaptscale age  
0           NaN  17  
1           NaN  17  
2           NaN  18  
3           NaN  18  
4           NaN  19  

[5 rows x 106 columns]

In [4]:
print(list(dataset.columns))

['outid', 'birth', 'sextypename', 'regname', 'areaname', 'tername', 'regtypename', 'eoname', 'eotypename', 'eoregname', 'eoareaname', 'eotername', 'eoparent', 'ukrtest', 'ukrteststatus', 'ukrptname', 'ukrptregname', 'ukrptareaname', 'ukrpttername', 'histtest', 'histlang', 'histteststatus', 'histptname', 'histptregname', 'histptareaname', 'histpttername', 'mathtest', 'mathlang', 'mathteststatus', 'mathptname', 'mathptregname', 'mathptareaname', 'mathpttername', 'phystest', 'physlang', 'physteststatus', 'physptname', 'physptregname', 'physptareaname', 'physpttername', 'chemtest', 'chemlang', 'chemteststatus', 'chemptname', 'chemptregname', 'chemptareaname', 'chempttername', 'biotest', 'biolang', 'bioteststatus', 'bioptname', 'bioptregname', 'bioptareaname', 'biopttername', 'geotest', 'geolang', 'geoteststatus', 'geoptname', 'geoptregname', 'geoptareaname', 'geopttername', 'engtest', 'engteststatus', 'engptname', 'engptregname', 'engptareaname', 'engpttername', 'deutest', 'deuteststatus',

In [5]:
dataset.shape

(1458122, 106)

# Feature Description

Dataset contains 1,458,122 individuals and 105 features (general and test performance)

## General features (demographic, educational etc.)

`outid` – unique identifier of the participant

`birth` – year of birth of the participant (1944-2007)

`sextypename` – gender of the participant ('Male' 'Female')

`regname` – region of registration/residence of the participant in Ukraine (24 oblasts and Kyiv, the capital of the country)

`areaname` – district/city of registration/residence of the participant

`tername` – place of registration/residence of the participant

`regtypename` – status of the participant (5 types: 'A graduate of an Ukrainian school of the current year', 'College graduate', 'A student of a higher education institution', 'A graduate of a school of the previous years', 'A graduate of a foreign school')

`tertypename` – type of territory (3 types: 'Village', 'City', 'Urban-type settlement')

`classprofilename`– profile of the participant's training class (24 types and two specific types: 'Junior specialit' for students and 'Skilled worker' for graduated people). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`classlangname` – participant's language of study (6 types and 'other' category). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoname` – the name of participant's institution of education. Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eotypename` – type of educational institution (30 types: 'Lyceum', 'Gymnasium', etc). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoregname` – region where the educational institution is located (same list as in regname). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoareaname` – district/city where the educational institution is located (same list as areaname). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eotername` – place where the educational institution is located (same list as tername). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoparent` – the authority to which the educational institution is subordinated. Full information for the graduates of a Ukrainian school of the current year (except 2021), partially for college graduates and students of a higher education institution. For else the values are NaN.

`year` - individual's participation year

`age` - individual's age at the moment of the participation

## Test performance features 

There are several subjects to choose from:
- ukr (ukrainian language and literature)
- hist (Ukrainian history)
- math (Mathematics)
- phys (Physics)
- chem (Chemistry)
- bio (biology)
- geo (geography)
- eng (English language)
- fr (French language)
- deu (Deutch language)
- sp (Spanish language)

We will decode one of them as `subj` instead of one of the 11 subjects mentioned above.

So, for each subject we have the following features:

`subjtest` - the name of the subject (corresponding name or nan if the participant didn't choose to take it)

`subjteststatus` – the result of the test ('The score threshold wasn't reached', 'Test passed', 'No attendance',
 'Test results were cancelled', '100-200 wasn't selected', or NaN if the participant didn't choose to take it)
       
`subjball`– raw score of the corresponding subject for participants with statuses: 'Test passed' and 'The score threshold wasn't reached' by subjteststatus (for else is NaN).

`subjptname`– the place of conducting the ZNO (corresponding name or nan if the participant didn't choose to take it)

`subjptregname` – the region where the ZNO is conducted (corresponding name, similar to regname, or nan if the participant didn't choose to take it)

`subjptareaname` – district/city where the ZNO testing center is located (corresponding name, similar to areaname, or nan if the participant didn't choose to take it)

`subjpttername` – place where the ZNO testing center is located (corresponding name, similar to tername, or nan if the participant didn't choose to take it)

In [9]:
for cols in dataset.columns:
    print(cols, ':', end=' ')
    if len(dataset[cols].unique())>6:
        print('{ number:', len(dataset[cols].unique()), ', nan:', dataset[cols].isnull().sum(),'}')
    else:
        print(dataset[cols].unique(), ', nan:', dataset[cols].isnull().sum())

outid : { number: 1458122 , nan: 0 }
birth : { number: 58 , nan: 0 }
sextypename : ['Male' 'Female'] , nan: 0
regname : { number: 25 , nan: 0 }
areaname : { number: 561 , nan: 0 }
tername : { number: 9366 , nan: 0 }
regtypename : ['A graduate of an Ukrainian school of the current year'
 'College graduate' 'A student of a higher education institution'
 'A graduate of a school of the previous years'
 'A graduate of a foreign school'] , nan: 0
eoname : { number: 18016 , nan: 160630 }
eotypename : { number: 29 , nan: 160630 }
eoregname : { number: 26 , nan: 160630 }
eoareaname : { number: 686 , nan: 160630 }
eotername : { number: 5447 , nan: 160630 }
eoparent : { number: 2508 , nan: 164292 }
ukrtest : ['Ukrainian language and literature' nan] , nan: 155349
ukrteststatus : ["The score threshold wasn't reached" 'Test passed' 'No attendance' nan
 'Test results were cancelled' "100-200 wasn't selected"] , nan: 155349
ukrptname : { number: 2388 , nan: 155349 }
ukrptregname : { number: 26 , nan:

In [8]:
dataset['eotypename'].value_counts()

eotypename
Secondary general education school                                443436
Higher educational institution                                    185117
Educational complex                                               105778
Lyceum                                                            103975
Higher educational institution of I-II levels of accreditation     75987
Gymnasium                                                          64320
Specialized school                                                 63684
Professional lyceum of the relevant profile                        48754
Institution of vocational education                                48714
Institution of professional preliminary higher education           46341
Higher vocational school                                           41481
Vocational school of the relevant profile                          24808
Center of professional and technical education                     12987
Educational association                 

In [16]:
dataset[dataset['classlangname'].isnull()].regtypename.unique()

array(['College graduate', 'A student of a higher education institution',
       'A graduate of a school of the previous years',
       'A graduate of a foreign school'], dtype=object)

In [15]:
dataset[dataset['classlangname'].notnull()].regtypename.unique()

array(['A graduate of an Ukrainian school of the current year',
       'A student of a higher education institution', 'College graduate'],
      dtype=object)

In [14]:
dataset['classprofilename'].unique()

array(['Physics and mathematics', 'Foreign philology', nan, 'Universal',
       'Ukrainian philology', 'Technology', 'Biology and physics',
       'Information technology', 'Biotechnology', 'Biology and chemistry',
       'Economics', 'History', 'Sports', 'Mathematics', 'Law',
       'Other (multi-profile)', 'Military and sports', 'Philosophy',
       'Ecology', 'Art and aesthetics', 'Geography', 'Physics',
       'Chemical technologies and agrochemistry', 'Biology',
       'Physics and chemistry', 'Junior specialist', 'Skilled worker'],
      dtype=object)

In [11]:
dataset.classprofilename.value_counts()

classprofilename
Junior specialist                          227506
Ukrainian philology                        187661
Universal                                  180839
Skilled worker                             147912
Foreign philology                           92907
Mathematics                                 65655
History                                     49086
Technology                                  42557
Other (multi-profile)                       37815
Information technology                      28526
Biology and chemistry                       18513
Physics and mathematics                     17838
Economics                                   16904
Law                                         14755
Military and sports                          9678
Sports                                       9122
Biology                                      6953
Biotechnology                                6471
Ecology                                      5601
Geography                        

In [10]:
dataset.classlangname.value_counts()

classlangname
Ukrainian    1117557
Russian        53948
Romanian        3778
Hungarian       2912
Moldovan         633
Polish           366
Other             91
Name: count, dtype: int64